File: sdks/python/docflow/storage/local.py
"""Local filesystem storage backend."""

import os
from pathlib import Path
from typing import BinaryIO, List, Optional

from .base import Storage


class LocalStorage(Storage):
    """Local filesystem storage implementation.
    
    Example:
        >>> storage = LocalStorage("./output")
        >>> storage.save("doc.pdf", pdf_bytes)
        >>> exists = storage.exists("doc.pdf")
    """

    def __init__(self, base_path: str) -> None:
        """Create local storage at the given path.
        
        Args:
            base_path: Root directory for all operations.
        """
        self.base_path = Path(base_path).resolve()
        self.base_path.mkdir(parents=True, exist_ok=True)

    def save(self, path: str, data: bytes) -> None:
        """Save data to the given path."""
        full_path = self._full_path(path)
        full_path.parent.mkdir(parents=True, exist_ok=True)
        full_path.write_bytes(data)

    def save_file(self, path: str, file: BinaryIO) -> None:
        """Save data from a file-like object."""
        full_path = self._full_path(path)
        full_path.parent.mkdir(parents=True, exist_ok=True)
        with open(full_path, "wb") as f:
            while chunk := file.read(8192):
                f.write(chunk)

    def load(self, path: str) -> bytes:
        """Load data from the given path."""
        full_path = self._full_path(path)
        if not full_path.exists():
            raise FileNotFoundError(f"File not found: {path}")
        return full_path.read_bytes()

    def delete(self, path: str) -> None:
        """Delete file at the given path."""
        full_path = self._full_path(path)
        if full_path.exists():
            full_path.unlink()

    def exists(self, path: str) -> bool:
        """Check if file exists at the given path."""
        return self._full_path(path).exists()

    def list(self, directory: str = "") -> List[str]:
        """List files in the given directory."""
        dir_path = self._full_path(directory)
        if not dir_path.exists():
            return []
        return [f.name for f in dir_path.iterdir() if f.is_file()]

    def get_url(self, path: str) -> Optional[str]:
        """Get file:// URL for the file."""
        return f"file://{self._full_path(path)}"

    def get_absolute_path(self, path: str) -> str:
        """Get the absolute filesystem path."""
        return str(self._full_path(path))

    def _full_path(self, path: str) -> Path:
        """Get full path for a relative path."""
        return self.base_path / path

File: sdks/go/docflow/storage/local.go
package storage

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
)

// LocalStorage implements Storage interface for local filesystem.
type LocalStorage struct {
	basePath string
}

// NewLocalStorage creates a new local storage instance.
// basePath is the root directory for all operations.
func NewLocalStorage(basePath string) (*LocalStorage, error) {
	// Create base directory if it doesn't exist
	if err := os.MkdirAll(basePath, 0755); err != nil {
		return nil, fmt.Errorf("failed to create base directory: %w", err)
	}

	absPath, err := filepath.Abs(basePath)
	if err != nil {
		return nil, fmt.Errorf("failed to get absolute path: %w", err)
	}

	return &LocalStorage{basePath: absPath}, nil
}

// Save stores data at the given path.
func (s *LocalStorage) Save(path string, data []byte) error {
	fullPath := s.fullPath(path)

	// Ensure parent directory exists
	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
		return fmt.Errorf("failed to create directory: %w", err)
	}

	if err := os.WriteFile(fullPath, data, 0644); err != nil {
		return fmt.Errorf("failed to write file: %w", err)
	}

	return nil
}

// SaveReader stores data from a reader at the given path.
func (s *LocalStorage) SaveReader(path string, reader io.Reader) error {
	fullPath := s.fullPath(path)

	// Ensure parent directory exists
	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
		return fmt.Errorf("failed to create directory: %w", err)
	}

	file, err := os.Create(fullPath)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer file.Close()

	if _, err := io.Copy(file, reader); err != nil {
		return fmt.Errorf("failed to write file: %w", err)
	}

	return nil
}

// Load retrieves data from the given path.
func (s *LocalStorage) Load(path string) ([]byte, error) {
	fullPath := s.fullPath(path)

	data, err := os.ReadFile(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("file not found: %s", path)
		}
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	return data, nil
}

// LoadReader returns a reader for the data at the given path.
func (s *LocalStorage) LoadReader(path string) (io.ReadCloser, error) {
	fullPath := s.fullPath(path)

	file, err := os.Open(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("file not found: %s", path)
		}
		return nil, fmt.Errorf("failed to open file: %w", err)
	}

	return file, nil
}

// Delete removes the file at the given path.
func (s *LocalStorage) Delete(path string) error {
	fullPath := s.fullPath(path)

	if err := os.Remove(fullPath); err != nil {
		if os.IsNotExist(err) {
			return nil // Already deleted
		}
		return fmt.Errorf("failed to delete file: %w", err)
	}

	return nil
}

// Exists checks if a file exists at the given path.
func (s *LocalStorage) Exists(path string) (bool, error) {
	fullPath := s.fullPath(path)

	_, err := os.Stat(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, fmt.Errorf("failed to check file: %w", err)
	}

	return true, nil
}

// List returns all files in the given directory.
func (s *LocalStorage) List(dir string) ([]string, error) {
	fullPath := s.fullPath(dir)

	entries, err := os.ReadDir(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return []string{}, nil
		}
		return nil, fmt.Errorf("failed to list directory: %w", err)
	}

	var files []string
	for _, entry := range entries {
		if !entry.IsDir() {
			files = append(files, entry.Name())
		}
	}

	return files, nil
}

// GetURL returns the file path as a file:// URL.
func (s *LocalStorage) GetURL(path string) string {
	return "file://" + s.fullPath(path)
}

// BasePath returns the base path of the storage.
func (s *LocalStorage) BasePath() string {
	return s.basePath
}

func (s *LocalStorage) fullPath(path string) string {
	return filepath.Join(s.basePath, path)
}
